File: Year-2023/LeetCode-July-Challenge/LongestArithmeticSubsequenceofGivenDifference.java
import java.util.*;

class Solution {
    public int longestSubsequence(int[] arr, int difference) {
        HashMap<Integer, Integer> mp = new HashMap<>();
        int res = 1;

        for (int x : arr) {
            if (mp.containsKey(x - difference))
                mp.put(x, mp.get(x - difference) + 1);
            else
                mp.put(x, 1);
            res = Math.max(res, mp.get(x));
        }
        return res;
    }
}

public class LongestArithmeticSubsequenceofGivenDifference {
    public static void main(String[] args) {
        int arr[] = { 1, 2, 3, 4 };
        int difference = 1;

        System.out.println(new Solution().longestSubsequence(arr, difference));
    }
}

File: Year-2023/LeetCode-July-Challenge/LongestArithmeticSubsequenceofGivenDifference.js
var longestSubsequence = function (arr, difference) {
    let mp = {}, res = 1;

    for (let x of arr) {
        if (mp[x - difference])
            mp[x] = mp[x - difference] + 1;
        else mp[x] = 1;
        res = Math.max(res, mp[x]);
    }

    return res;
};

let arr = [1, 2, 3, 4], difference = 1;
console.log(longestSubsequence(arr, difference));